File: TypeScript/Algorithm/Algorithm I/278. First Bad Version.ts
/**
 * Represents a solution to find the first bad version using binary search.
 * @param {function} isBadVersion - A function that determines if a given version is bad.
 * @returns {function} A function that returns the first bad version.
 */
const solution = function (isBadVersion: any): Function {
  return function (n: number): number {
    let leftPointer: number = 1;
    let rightPointer: number = n;

    while (leftPointer <= rightPointer) {
      const middlePointer: number = Math.floor((leftPointer + rightPointer) / 2);

      if (isBadVersion(middlePointer)) rightPointer = middlePointer - 1;
      else leftPointer = middlePointer + 1;
    }

    return rightPointer + 1;
  };
};

File: Java/Algorithm/Algorithm I/278. First Bad Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
   boolean isBadVersion(int version); */

public class Solution extends VersionControl {

  /**
   * Finds the first bad version in a range of versions using binary search.
   *
   * @param n The highest version number.
   * @return The first bad version found.
   */
  public int firstBadVersion(int n) {
    int leftPointer = 1;
    int rightPointer = n;

    while (leftPointer <= rightPointer) {
      int middlePointer = leftPointer + (rightPointer - leftPointer) / 2;

      if (isBadVersion(middlePointer)) 
        rightPointer = middlePointer - 1;
      
      else 
        leftPointer = middlePointer + 1;
    }

    return rightPointer + 1;
  }
}
